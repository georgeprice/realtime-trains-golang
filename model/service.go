package model

import (
	"reflect"
)

// ServiceType describes what type of train is running
type ServiceType string

// BusService is a bus?
// ShipService is a ship?
// TrainService is a train? (Maybe they're all trains, idk)
const (
	BusService   ServiceType = "bus"
	ShipService  ServiceType = "ship"
	TrainService ServiceType = "train"
)

// Service defines the response from a request for a particular train running between two points
type Service struct {
	ServiceUID           string `json:"serviceUid,omitempty"`
	RunDate              string `json:"runDate,omitempty"`
	ServiceType          `json:"serviceType,omitempty"`
	IsPassenger          bool             `json:"isPassenger,omitempty"`
	TrainIdentity        string           `json:"trainIdentity,omitempty"`
	PowerType            string           `json:"powerType,omitempty"`
	TrainClass           string           `json:"trainClass,omitempty"`
	Sleeper              string           `json:"sleeper,omitempty"`
	ATOCCode             string           `json:"atocCode,omitempty"`
	ATOCName             string           `json:"atocName,omitempty"`
	PerformanceMonitored bool             `json:"performanceMonitored,omitempty"`
	Origin               []Pair           `json:"origin,omitempty"`
	Destination          []Pair           `json:"destination,omitempty"`
	Locations            []LocationDetail `json:"locations,omitempty"`
	RealtimeActivated    bool             `json:"realtimeActivated,omitempty"`
	RunningIdentity      string           `json:"runningIdentity,omitempty"`
	PlannedCancel        bool             `json:"plannedCancel,omitempty"`
}

// FresherThan returns whether another service struct contains more up-to-date data
//	I just assume that different data is newer data lol
func (s Service) FresherThan(thou Service) bool {
	return !reflect.DeepEqual(s, thou)
}

// Pair describes a start or end of a train's journey (don't ask)
type Pair struct {
	TIPLOC      string `json:"tiploc,omitempty"`
	Description string `json:"description,omitempty"`
	WorkingTime string `json:"workingTime,omitempty"`
	PublicTime  string `json:"publicTime,omitempty"`
}

// LocationDetail describes a station which is passed through between the origin and destination of a service
type LocationDetail struct {
	RealTimeActivated bool   `json:"realtimeActivated,omitempty"`
	TIPLOC            string `json:"tiploc,omitempty"`
	CRS               string `json:"crs,omitempty"`
	Description       string `json:"description,omitempty"`

	WTTBookedArrival           string `json:"wttBookedArrival,omitempty"`
	WTTBookedDeparture         string `json:"wttBookedDeparture,omitempty"`
	WTTBookedPass              string `json:"wttBookedPass,omitempty"`
	GBTTBookedArrival          string `json:"gbttBookedArrival,omitempty"`
	GBTTBookedArrivalNextDay   bool   `json:"gbttBookedArrivalNextDay,omitempty"`
	GBTTBookedDeparture        string `json:"gbttBookedDeparture,omitempty"`
	GBTTBookedDepartureNextDay bool   `json:"gbttBookedDepartureNextDay,omitempty"`

	Origin       []Pair `json:"origin,omitempty"`
	Destination  []Pair `json:"destination,omitempty"`
	IsCall       bool   `json:"isCall,omitempty"`
	IsCallPublic bool   `json:"isPublicCall,omitempty"`

	RealTimeArrival         string `json:"realtimeArrival,omitempty"`
	RealTimeArrivalActual   bool   `json:"realtimeArrivalActual,omitempty"`
	RealTimeArrivalNoReport bool   `json:"realtimeArrivalNoReport,omitempty"`
	RealTimeArrivalNextDay  bool   `json:"realtimeArrivalNextDay,omitempty"`

	RealTimeGBTTArrivalLateness        int `json:"realtimeGbttArrivalLateness,omitempty"`
	RealTimeWTTArrivalLateness         int `json:"realtimeWttDepartureLateness,omitempty"`
	RealTimeWTTArrivalLatenessDetailed int `json:"realtimeWttArrivalLateness,omitempty"`

	RealTimeDeparture         string `json:"realtimeDeparture,omitempty"`
	RealTimeDepartureActual   bool   `json:"realtimeDepartureActual,omitempty"`
	RealTimeDepartureNoReport bool   `json:"realtimeDepartureNoReport,omitempty"`
	RealTimeDepartureNextDay  bool   `json:"realtimeDepartureNextDay,omitempty"`

	RealTimePass         string `json:"realtimePass,omitempty"`
	RealTimePassActual   bool   `json:"realtimePassActual,omitempty"`
	RealTimePassNoReport bool   `json:"realtimePassNoReport,omitempty"`

	Platform              string `json:"platform,omitempty"`
	PlatformConfirmed     bool   `json:"platformConfirmed,omitempty"`
	PlatformChanged       bool   `json:"platformChanged,omitempty"`
	Line                  string `json:"line,omitempty"`
	LineConfirmed         bool   `json:"lineConfirmed,omitempty"`
	Path                  string `json:"path,omitempty"`
	PathConfirmed         bool   `json:"pathConfirmed,omitempty"`
	CancelReasonCode      string `json:"cancelReasonCode,omitempty"`
	CancelReasonShortText string `json:"cancelReasonShortText,omitempty"`
	CancelReasonLongText  string `json:"cancelReasonLongText,omitempty"`
	DisplayAs             string `json:"displayAs,omitempty"`
	ServiceLocation       string `json:"serviceLocation,omitempty"`
}
