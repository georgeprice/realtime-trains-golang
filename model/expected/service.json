{
  "serviceUid": "Q13773",
  "runDate": "2020-02-12",
  "serviceType": "train",
  "isPassenger": true,
  "trainIdentity": "1Z73",
  "powerType": "EMU",
  "trainClass": "B",
  "atocCode": "SW",
  "atocName": "South Western Railway",
  "performanceMonitored": true,
  "origin": [
    {
      "tiploc": "ELGH",
      "description": "Eastleigh",
      "workingTime": "233700",
      "publicTime": "2337"
    }
  ],
  "destination": [
    {
      "tiploc": "POOLE",
      "description": "Poole",
      "workingTime": "004700",
      "publicTime": "0047"
    }
  ],
  "locations": [
    {
      "realtimeActivated": true,
      "tiploc": "ELGH",
      "crs": "ESL",
      "description": "Eastleigh",
      "gbttBookedDeparture": "2337",
      "origin": [
        {
          "tiploc": "ELGH",
          "description": "Eastleigh",
          "workingTime": "233700",
          "publicTime": "2337"
        }
      ],
      "destination": [
        {
          "tiploc": "POOLE",
          "description": "Poole",
          "workingTime": "004700",
          "publicTime": "0047"
        }
      ],
      "isCall": true,
      "isPublicCall": true,
      "realtimeDeparture": "2337",
      "realtimeDepartureActual": false,
      "platform": "2N",
      "platformConfirmed": false,
      "platformChanged": false,
      "displayAs": "ORIGIN"
    },
    {
      "realtimeActivated": true,
      "tiploc": "SOTPKWY",
      "crs": "SOA",
      "description": "Southampton Airport Parkway",
      "gbttBookedArrival": "2340",
      "gbttBookedDeparture": "2341",
      "origin": [
        {
          "tiploc": "ELGH",
          "description": "Eastleigh",
          "workingTime": "233700",
          "publicTime": "2337"
        }
      ],
      "destination": [
        {
          "tiploc": "POOLE",
          "description": "Poole",
          "workingTime": "004700",
          "publicTime": "0047"
        }
      ],
      "isCall": true,
      "isPublicCall": true,
      "realtimeArrival": "2340",
      "realtimeArrivalActual": false,
      "realtimeDeparture": "2341",
      "realtimeDepartureActual": false,
      "platform": "2",
      "platformConfirmed": false,
      "platformChanged": false,
      "displayAs": "CALL"
    },
    {
      "realtimeActivated": true,
      "tiploc": "SOTON",
      "crs": "SOU",
      "description": "Southampton Central",
      "gbttBookedArrival": "2348",
      "gbttBookedDeparture": "2350",
      "origin": [
        {
          "tiploc": "ELGH",
          "description": "Eastleigh",
          "workingTime": "233700",
          "publicTime": "2337"
        }
      ],
      "destination": [
        {
          "tiploc": "POOLE",
          "description": "Poole",
          "workingTime": "004700",
          "publicTime": "0047"
        }
      ],
      "isCall": true,
      "isPublicCall": true,
      "realtimeArrival": "2347",
      "realtimeArrivalActual": false,
      "realtimeDeparture": "2350",
      "realtimeDepartureActual": false,
      "platform": "4",
      "platformConfirmed": false,
      "platformChanged": false,
      "line": "FL",
      "lineConfirmed": false,
      "displayAs": "CALL"
    },
    {
      "realtimeActivated": true,
      "tiploc": "TOTTON",
      "crs": "TTN",
      "description": "Totton",
      "gbttBookedArrival": "2355",
      "gbttBookedDeparture": "2355",
      "origin": [
        {
          "tiploc": "ELGH",
          "description": "Eastleigh",
          "workingTime": "233700",
          "publicTime": "2337"
        }
      ],
      "destination": [
        {
          "tiploc": "POOLE",
          "description": "Poole",
          "workingTime": "004700",
          "publicTime": "0047"
        }
      ],
      "isCall": true,
      "isPublicCall": true,
      "realtimeArrival": "2354",
      "realtimeArrivalActual": false,
      "realtimeDeparture": "2355",
      "realtimeDepartureActual": false,
      "platform": "2",
      "platformConfirmed": false,
      "platformChanged": false,
      "displayAs": "CALL"
    },
    {
      "realtimeActivated": true,
      "tiploc": "LYNDHRD",
      "crs": "ANF",
      "description": "Ashurst New Forest",
      "gbttBookedArrival": "2359",
      "gbttBookedDeparture": "2359",
      "origin": [
        {
          "tiploc": "ELGH",
          "description": "Eastleigh",
          "workingTime": "233700",
          "publicTime": "2337"
        }
      ],
      "destination": [
        {
          "tiploc": "POOLE",
          "description": "Poole",
          "workingTime": "004700",
          "publicTime": "0047"
        }
      ],
      "isCall": true,
      "isPublicCall": true,
      "realtimeArrival": "2359",
      "realtimeArrivalActual": false,
      "realtimeDeparture": "2359",
      "realtimeDepartureActual": false,
      "displayAs": "CALL"
    },
    {
      "realtimeActivated": true,
      "tiploc": "BKNHRST",
      "crs": "BCU",
      "description": "Brockenhurst",
      "gbttBookedArrival": "0008",
      "gbttBookedArrivalNextDay": true,
      "gbttBookedDeparture": "0009",
      "gbttBookedDepartureNextDay": true,
      "origin": [
        {
          "tiploc": "ELGH",
          "description": "Eastleigh",
          "workingTime": "233700",
          "publicTime": "2337"
        }
      ],
      "destination": [
        {
          "tiploc": "POOLE",
          "description": "Poole",
          "workingTime": "004700",
          "publicTime": "0047"
        }
      ],
      "isCall": true,
      "isPublicCall": true,
      "realtimeArrival": "0007",
      "realtimeArrivalActual": false,
      "realtimeArrivalNextDay": true,
      "realtimeDeparture": "0009",
      "realtimeDepartureActual": false,
      "realtimeDepartureNextDay": true,
      "platform": "3",
      "platformConfirmed": false,
      "platformChanged": false,
      "displayAs": "CALL"
    },
    {
      "realtimeActivated": true,
      "tiploc": "SWAY",
      "crs": "SWY",
      "description": "Sway",
      "gbttBookedArrival": "0013",
      "gbttBookedArrivalNextDay": true,
      "gbttBookedDeparture": "0013",
      "gbttBookedDepartureNextDay": true,
      "origin": [
        {
          "tiploc": "ELGH",
          "description": "Eastleigh",
          "workingTime": "233700",
          "publicTime": "2337"
        }
      ],
      "destination": [
        {
          "tiploc": "POOLE",
          "description": "Poole",
          "workingTime": "004700",
          "publicTime": "0047"
        }
      ],
      "isCall": true,
      "isPublicCall": true,
      "realtimeArrival": "0013",
      "realtimeArrivalActual": false,
      "realtimeArrivalNextDay": true,
      "realtimeDeparture": "0013",
      "realtimeDepartureActual": false,
      "realtimeDepartureNextDay": true,
      "displayAs": "CALL"
    },
    {
      "realtimeActivated": true,
      "tiploc": "NMILTON",
      "crs": "NWM",
      "description": "New Milton",
      "gbttBookedArrival": "0018",
      "gbttBookedArrivalNextDay": true,
      "gbttBookedDeparture": "0018",
      "gbttBookedDepartureNextDay": true,
      "origin": [
        {
          "tiploc": "ELGH",
          "description": "Eastleigh",
          "workingTime": "233700",
          "publicTime": "2337"
        }
      ],
      "destination": [
        {
          "tiploc": "POOLE",
          "description": "Poole",
          "workingTime": "004700",
          "publicTime": "0047"
        }
      ],
      "isCall": true,
      "isPublicCall": true,
      "realtimeArrival": "0017",
      "realtimeArrivalActual": false,
      "realtimeArrivalNextDay": true,
      "realtimeDeparture": "0018",
      "realtimeDepartureActual": false,
      "realtimeDepartureNextDay": true,
      "displayAs": "CALL"
    },
    {
      "realtimeActivated": true,
      "tiploc": "HINTONA",
      "crs": "HNA",
      "description": "Hinton Admiral",
      "gbttBookedArrival": "0022",
      "gbttBookedArrivalNextDay": true,
      "gbttBookedDeparture": "0022",
      "gbttBookedDepartureNextDay": true,
      "origin": [
        {
          "tiploc": "ELGH",
          "description": "Eastleigh",
          "workingTime": "233700",
          "publicTime": "2337"
        }
      ],
      "destination": [
        {
          "tiploc": "POOLE",
          "description": "Poole",
          "workingTime": "004700",
          "publicTime": "0047"
        }
      ],
      "isCall": true,
      "isPublicCall": true,
      "realtimeArrival": "0021",
      "realtimeArrivalActual": false,
      "realtimeArrivalNextDay": true,
      "realtimeDeparture": "0022",
      "realtimeDepartureActual": false,
      "realtimeDepartureNextDay": true,
      "displayAs": "CALL"
    },
    {
      "realtimeActivated": true,
      "tiploc": "CHRISTC",
      "crs": "CHR",
      "description": "Christchurch",
      "gbttBookedArrival": "0026",
      "gbttBookedArrivalNextDay": true,
      "gbttBookedDeparture": "0027",
      "gbttBookedDepartureNextDay": true,
      "origin": [
        {
          "tiploc": "ELGH",
          "description": "Eastleigh",
          "workingTime": "233700",
          "publicTime": "2337"
        }
      ],
      "destination": [
        {
          "tiploc": "POOLE",
          "description": "Poole",
          "workingTime": "004700",
          "publicTime": "0047"
        }
      ],
      "isCall": true,
      "isPublicCall": true,
      "realtimeArrival": "0026",
      "realtimeArrivalActual": false,
      "realtimeArrivalNextDay": true,
      "realtimeDeparture": "0027",
      "realtimeDepartureActual": false,
      "realtimeDepartureNextDay": true,
      "displayAs": "CALL"
    },
    {
      "realtimeActivated": true,
      "tiploc": "POKSDWN",
      "crs": "POK",
      "description": "Pokesdown",
      "gbttBookedArrival": "0030",
      "gbttBookedArrivalNextDay": true,
      "gbttBookedDeparture": "0030",
      "gbttBookedDepartureNextDay": true,
      "origin": [
        {
          "tiploc": "ELGH",
          "description": "Eastleigh",
          "workingTime": "233700",
          "publicTime": "2337"
        }
      ],
      "destination": [
        {
          "tiploc": "POOLE",
          "description": "Poole",
          "workingTime": "004700",
          "publicTime": "0047"
        }
      ],
      "isCall": true,
      "isPublicCall": true,
      "realtimeArrival": "0030",
      "realtimeArrivalActual": false,
      "realtimeArrivalNextDay": true,
      "realtimeDeparture": "0030",
      "realtimeDepartureActual": false,
      "realtimeDepartureNextDay": true,
      "displayAs": "CALL"
    },
    {
      "realtimeActivated": true,
      "tiploc": "BOMO",
      "crs": "BMH",
      "description": "Bournemouth",
      "gbttBookedArrival": "0034",
      "gbttBookedArrivalNextDay": true,
      "gbttBookedDeparture": "0036",
      "gbttBookedDepartureNextDay": true,
      "origin": [
        {
          "tiploc": "ELGH",
          "description": "Eastleigh",
          "workingTime": "233700",
          "publicTime": "2337"
        }
      ],
      "destination": [
        {
          "tiploc": "POOLE",
          "description": "Poole",
          "workingTime": "004700",
          "publicTime": "0047"
        }
      ],
      "isCall": true,
      "isPublicCall": true,
      "realtimeArrival": "0034",
      "realtimeArrivalActual": false,
      "realtimeArrivalNextDay": true,
      "realtimeDeparture": "0036",
      "realtimeDepartureActual": false,
      "realtimeDepartureNextDay": true,
      "platform": "3",
      "platformConfirmed": false,
      "platformChanged": false,
      "displayAs": "CALL"
    },
    {
      "realtimeActivated": true,
      "tiploc": "BRANKSM",
      "crs": "BSM",
      "description": "Branksome",
      "gbttBookedArrival": "0041",
      "gbttBookedArrivalNextDay": true,
      "gbttBookedDeparture": "0041",
      "gbttBookedDepartureNextDay": true,
      "origin": [
        {
          "tiploc": "ELGH",
          "description": "Eastleigh",
          "workingTime": "233700",
          "publicTime": "2337"
        }
      ],
      "destination": [
        {
          "tiploc": "POOLE",
          "description": "Poole",
          "workingTime": "004700",
          "publicTime": "0047"
        }
      ],
      "isCall": true,
      "isPublicCall": true,
      "realtimeArrival": "0040",
      "realtimeArrivalActual": false,
      "realtimeArrivalNextDay": true,
      "realtimeDeparture": "0041",
      "realtimeDepartureActual": false,
      "realtimeDepartureNextDay": true,
      "platform": "2",
      "platformConfirmed": false,
      "platformChanged": false,
      "displayAs": "CALL"
    },
    {
      "realtimeActivated": true,
      "tiploc": "PSTONE",
      "crs": "PKS",
      "description": "Parkstone (Dorset)",
      "gbttBookedArrival": "0044",
      "gbttBookedArrivalNextDay": true,
      "gbttBookedDeparture": "0044",
      "gbttBookedDepartureNextDay": true,
      "origin": [
        {
          "tiploc": "ELGH",
          "description": "Eastleigh",
          "workingTime": "233700",
          "publicTime": "2337"
        }
      ],
      "destination": [
        {
          "tiploc": "POOLE",
          "description": "Poole",
          "workingTime": "004700",
          "publicTime": "0047"
        }
      ],
      "isCall": true,
      "isPublicCall": true,
      "realtimeArrival": "0043",
      "realtimeArrivalActual": false,
      "realtimeArrivalNextDay": true,
      "realtimeDeparture": "0044",
      "realtimeDepartureActual": false,
      "realtimeDepartureNextDay": true,
      "displayAs": "CALL"
    },
    {
      "realtimeActivated": true,
      "tiploc": "POOLE",
      "crs": "POO",
      "description": "Poole",
      "gbttBookedArrival": "0047",
      "gbttBookedArrivalNextDay": true,
      "origin": [
        {
          "tiploc": "ELGH",
          "description": "Eastleigh",
          "workingTime": "233700",
          "publicTime": "2337"
        }
      ],
      "destination": [
        {
          "tiploc": "POOLE",
          "description": "Poole",
          "workingTime": "004700",
          "publicTime": "0047"
        }
      ],
      "isCall": true,
      "isPublicCall": true,
      "realtimeArrival": "0047",
      "realtimeArrivalActual": false,
      "realtimeArrivalNextDay": true,
      "platform": "1",
      "platformConfirmed": false,
      "platformChanged": false,
      "displayAs": "DESTINATION"
    }
  ],
  "realtimeActivated": true,
  "runningIdentity": "1Z73",
  "plannedCancel": true
}